### PR TITLE
UDP BBA Emulation

### PR DESCRIPTION
**Pull request by tommy1019**
This PR adds a new option for BBA emulation to send packets out over a UDP socket as opposed to a TAP network.

Screen Shot 2020-04-19 at 7 20 04 PM

This allows for running BBA emulation without additional programs for TAP support.

The code was tested on two macOS machines running Mario Kart Double Dash on LAN Mode as shown below.

Screen Shot 2020-04-19 at 7 14 22 PM

Screen Shot 2020-04-19 at 7 14 58 PM

The UDP BBA emulation should also work on unix, as the code is the same. However I have not implemented the windows version of the UDP BBA emulation as I do not have access to a windows machine.

### CODE REVIEW

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 1]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
+    virtual void RecvStart(){};
+    virtual void RecvStop(){};
+  };
 
-  std::unique_ptr<u8[]> mRecvBuffer;
-  u32 mRecvBufferLength = 0;
+  class TAPPhysicalNetworkInterface : public PhysicalNetworkInterface
+  {
+  public:
+    TAPPhysicalNetworkInterface(CEXIETHERNET* ethRef)
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:318]:
```suggestion
    explicit TAPPhysicalNetworkInterface(CEXIETHERNET* eth_ref) : PhysicalNetworkInterface(eth_ref) {}
```

with the previous suggestion of supplying a protected constructor.

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 2]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
 #elif defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__)
-  int fd = -1;
+    int fd = -1;
 #endif
+  };
 
+  class UDPPhysicalNetworkInterface : public PhysicalNetworkInterface
+  {
+  public:
+    UDPPhysicalNetworkInterface(CEXIETHERNET* ethRef, std::string destIp, int destPort, int inPort)
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:354]:
```suggestion
    UDPPhysicalNetworkInterface(CEXIETHERNET* eth_ref, std::string dest_ip, int dest_port, int in_port)
```

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 3]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
+        : destIp(destIp), destPort(destPort), inPort(inPort)
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:355]:
```suggestion
        : destIp(std::move(destIp)), destPort(destPort), inPort(inPort)
```

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 4]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
+    {
+      PhysicalNetworkInterface::ethRef = ethRef;
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:357]: The forwarding constructor can be used here as well.

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 5]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
+    };
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:358]:
```suggestion
    }
```

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.cpp [hunk 6]
```diff
@@ -23,8 +23,30 @@ namespace ExpansionInterface
...
+
+    int colonPos = destStr.find(':');
+    if (colonPos < 0)
+      colonPos = destStr.length() - 1;
+
+    std::string ip = destStr.substr(0, colonPos);
+    int port = atoi(destStr.substr(colonPos + 1).c_str());
+
+    physical_network_interface = std::make_unique<UDPPhysicalNetworkInterface>(
+        this, ip, port, SConfig::GetInstance().m_bba_udp_port);
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.cpp:45]:
```suggestion
        this, std::move(ip), port, SConfig::GetInstance().m_bba_udp_port);
```

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.cpp [hunk 7]
```diff
@@ -23,8 +23,30 @@ namespace ExpansionInterface
...
+  if (tap)
+  {
+    physical_network_interface = std::make_unique<TAPPhysicalNetworkInterface>(this);
+    INFO_LOG(SP1, "Created TAP physical network interface.");
+  }
+  else
+  {
+    std::string destStr = SConfig::GetInstance().m_bba_udp_dest;
+
+    int colonPos = destStr.find(':');
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.cpp:37]:
```suggestion
    size_t colon_pos = destStr.find(':');
```

**Comment by JMC47** [comment]
I don't have a Linux Machine fast enough to really test this.  Is there anything in particular stopping this from working on Windows?  We've had other attempts to HLE BBA with UDP, but they've failed.

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.cpp [hunk 8]
```diff
@@ -23,8 +23,30 @@ namespace ExpansionInterface
...
+CEXIETHERNET::CEXIETHERNET(bool tap)
 {
+  if (tap)
+  {
+    physical_network_interface = std::make_unique<TAPPhysicalNetworkInterface>(this);
+    INFO_LOG(SP1, "Created TAP physical network interface.");
+  }
+  else
+  {
+    std::string destStr = SConfig::GetInstance().m_bba_udp_dest;
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.cpp:35]:
```suggestion
    const std::string& dest_str = SConfig::GetInstance().m_bba_udp_dest;
```

File: Source/Core/Core/HW/EXI/BBA/UDP_Unix.cpp [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+void CEXIETHERNET::UDPPhysicalNetworkInterface::ReadThreadHandler(
+    CEXIETHERNET::UDPPhysicalNetworkInterface* self)
+{
+  while (!self->readThreadShutdown.IsSet())
+  {
+    if (self->readFD < 0)
+      break;
+
+    int readBytes = read(self->readFD, self->ethRef->mRecvBuffer.get(), BBA_RECV_SIZE);
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/BBA/UDP_Unix.cpp:116]:
```suggestion
    const int read_bytes = read(self->readFD, self->ethRef->mRecvBuffer.get(), BBA_RECV_SIZE);
```

File: Source/Core/Core/HW/EXI/BBA/UDP_Unix.cpp [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+bool CEXIETHERNET::UDPPhysicalNetworkInterface::IsActivated()
+{
+  return readFD != -1 && writeFD == -1;
+}
+
+bool CEXIETHERNET::UDPPhysicalNetworkInterface::SendFrame(const u8* frame, u32 size)
+{
+  INFO_LOG(SP1, "SendFrame %x\n%s", size, ArrayToString(frame, size, 0x10).c_str());
+
+  int writtenBytes = write(writeFD, frame, size);
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/BBA/UDP_Unix.cpp:94]:
```suggestion
  const int written_bytes = write(writeFD, frame, size);
```

File: Source/Core/Core/HW/EXI/BBA/UDP_Unix.cpp [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+  if ((u32)writtenBytes != size)
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/BBA/UDP_Unix.cpp:95]:
```suggestion
  if (u32(writtenBytes) != size)
```

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 12]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
+    virtual bool Activate() override;
+    virtual void Deactivate() override;
+    virtual bool IsActivated() override;
+    virtual bool SendFrame(const u8* frame, u32 size) override;
+    virtual bool RecvInit() override;
+    virtual void RecvStart() override;
+    virtual void RecvStop() override;
+
+  private:
+    std::string destIp;
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:370]: Class variables in new code should be following the 

`m_name_of_thing` convention.

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 13]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
+  {
+  public:
+    UDPPhysicalNetworkInterface(CEXIETHERNET* ethRef, std::string destIp, int destPort, int inPort)
+        : destIp(destIp), destPort(destPort), inPort(inPort)
+    {
+      PhysicalNetworkInterface::ethRef = ethRef;
+    };
+
+  public:
+    virtual bool Activate() override;
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:361]: `virtual` can be omitted on all of these overrides.

File: Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h [hunk 14]
```diff
@@ -297,34 +297,97 @@ class CEXIETHERNET : public IEXIDevice
...
+  class TAPPhysicalNetworkInterface : public PhysicalNetworkInterface
+  {
+  public:
+    TAPPhysicalNetworkInterface(CEXIETHERNET* ethRef)
+    {
+      PhysicalNetworkInterface::ethRef = ethRef;
+    };
+
+  public:
+    virtual bool Activate() override;
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI/EXI_DeviceEthernet.h:324]: `virtual` can be omitted on these overrides.

**Comment by tommy1019** [comment]
> I don't have a Linux Machine fast enough to really test this. Is there anything in particular stopping this from working on Windows? We've had other attempts to HLE BBA in the past, but they've failed.

There is no reason it shouldn't work on windows, all the underlying data transfer should be the same. I just don't have a windows machine to write and test the winsock implementation.

**Comment by JMC47** [comment]
It was mentioned on IRC that this may only work with connecting one Dolphin instance to another?  A lot of BBA games (all of them?) support at least 4 consoles.  May be something to look into in the future.

**Comment by tommy1019** [comment]
I believe UDP multicast should be able to allow for more than 2 consoles connecting together, although this will require more testing, as well as a more complicated GUI and some kind of incoming packet filtering to ignore packets being repeated to the sender.

**Comment by CrunchBite** [comment]
Heya! Would you be interested in re-working this to use XLink Kai as the backend over UDP socket?
I've added a simple UDP based interface to the client that sends/receives encapsulated Ethernet frames to facilitate this.

**Comment by Tom2Drake** [comment]
Playing Games like Mario Kart double dash in LAN mode over the internet comes with some issues. The LAN mode is very sensitive to network latency. I have tested the server client architecture from this pull request https://github.com/dolphin-emu/dolphin/pull/7786 with a dedicated cloud server and the game client FPS dropped to 18 FPS with a latency of 14ms.

**Comment by JMC47** [comment]
Things like Xlink allow BBA games to be played at a much higher latency, making it more viable for online play.

**Comment by trevor403** [comment]
@tommy1019 are you still interested in getting this patch merged?

**Comment by Rumi-Larry** [comment]
Anybody interested in reviving this project?

**Comment by CrunchBite** [comment]
> Anybody interested in reviving this project?

This may not be needed anymore, see:

- https://github.com/dolphin-emu/dolphin/pull/8853
- https://dolphin-emu.org/blog/2020/07/05/dolphin-progress-report-may-and-june-2020/
- https://www.teamxlink.co.uk/wiki/Dolphin_Gamecube_XLink_BBA_Tutorial

**Comment by Miksel12** [comment]
I think an option to use BBA without external software would be nice to have.

**Comment by Rumi-Larry** [comment]
It may be wise to start with the refactoring https://github.com/dolphin-emu/dolphin/pull/7817

**Comment by sepalani** [comment]
This PR is probably obsoleted by https://dolp.in/pr10690.

**Comment by JMC47** [comment]
Yeah, this is definitely obsoleted by BBA HLE.  I still appreciate the work done for this back in the day though.

**Pull request closed without merging**